src/crimson/os/seastore/journal/circular_bounded_journal.cc

```diff
@@ -262,111 +366,50 @@ Journal::replay_ret CircularBoundedJournal::replay(
-	    get_device_id(), cursor_addr + r_header.mdlength);
-	  auto maybe_record_deltas_list = try_decode_deltas(
-	    r_header, mdbuf, record_block_base);
-	  if (!maybe_record_deltas_list) {
-	    // This should be impossible, we did check the crc on the mdbuf
-	    ERROR("unable to decode deltas for record {} at {}",
-		  r_header, record_block_base);
-	    return crimson::ct_error::input_output_error::make();
+	}
+	return replay_ertr::make_ready_future<bool>(true);
+      };
+      written_to.segment_seq = NULL_SEG_SEQ;
+      auto tail = get_dirty_tail() <= get_alloc_tail() ?
+	get_dirty_tail() : get_alloc_tail();
+      set_written_to(tail);
+      return for_each_records(std::move(add_invalid_records), tail
+      ).safe_then([this, &map, &d_handler, tail]() {
+	auto call_d_handler_if_valid = [this, &map, &d_handler](
+	  const auto &offsets,
+	  const auto &e,
+	  sea_time_point modify_time)
+	{
+	  if (map.find(e.paddr) == map.end() ||
+	      map[e.paddr] <= offsets.write_result.start_seq) {
+	    return d_handler(
+	      offsets,
+	      e,
+	      header.dirty_tail,
+	      header.alloc_tail,
+	      modify_time
+	    );
 	  }
-	  DEBUG("{} at {}", r_header, cursor_addr);
-	  auto write_result = write_result_t{
-	    r_header.committed_to,
-	    bl.length()
-	  };
-	  if (expected_seq == NULL_SEG_SEQ) {
-	    expected_seq = r_header.committed_to.segment_seq;
-	  } else {
-	    assert(expected_seq == r_header.committed_to.segment_seq);
-	  }
-	  cursor_addr += bl.length();
-	  if (cursor_addr >= get_journal_end()) {
-	    assert(cursor_addr == get_journal_end());
-	    cursor_addr = get_records_start();
-	    ++expected_seq;
-	    is_rolled = true;
-	  }
-	  paddr_t addr = convert_abs_addr_to_paddr(
-	    cursor_addr,
-	    get_device_id());
-	  set_written_to(
-	    journal_seq_t{expected_seq, addr});
-	  return seastar::do_with(
-	    std::move(*maybe_record_deltas_list),
-	    [this,
-	    write_result,
-	    &d_handler,
-	    FNAME](auto& record_deltas_list) {
-	    return crimson::do_for_each(
-	      record_deltas_list,
-	      [this,
-	      write_result,
-	      &d_handler, FNAME](record_deltas_t& record_deltas) {
-	      auto locator = record_locator_t{
-		record_deltas.record_block_base,
-		write_result
-	      };
-	      DEBUG("processing {} deltas at block_base {}",
-		  record_deltas.deltas.size(),
-		  locator);
-	      return crimson::do_for_each(
-		record_deltas.deltas,
-		[this,
-		locator,
-		&d_handler](auto& p) {
-		auto& modify_time = p.first;
-		auto& delta = p.second;
-		return d_handler(
-		  locator,
-		  delta,
-		  header.dirty_tail,
-		  header.alloc_tail,
-		  modify_time).discard_result();
-	      });
-	    }).safe_then([]() {
-	      return replay_ertr::make_ready_future<
-		seastar::stop_iteration>(seastar::stop_iteration::no);
-	    });
-	  });
-	});
+	  return replay_ertr::make_ready_future<bool>(true);
+	};
+	return for_each_records(std::move(call_d_handler_if_valid), tail);
```

Comment:
Add comments for the two passes before each `for_each_record_delta()`, probably:
```
// The first pass to build the paddr->journal_seq_t map according to extent allocations
// The second pass to replay deltas
```

Reply:
Done.